### PR TITLE
[AOTInductor] Include constants in AOTInductor .so file.

### PR DESCRIPTION
**Pull request by muchulee8**
Summary:
Include the constants into AOTInductor .so file.
We do not modify existing API signatures but create necessary format with weight lifted out instead.

Test Plan:
test/inductor/test_aot_inductor.py

Reviewers:

Subscribers:

Tasks:

Tags:

Fixes #ISSUE_NUMBER

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @aakhundov @anijain2305

Marking as already reverted: 
@diff-train-skip-merge
see https://github.com/pytorch/pytorch/commit/cf64a9e0a58f1252b63663b666ebc935eec616b2

### CODE REVIEW

File: torch/_inductor/compile_fx.py [hunk 1]
```diff
@@ -846,7 +849,7 @@ def compile_fx_aot(
     example_inputs_: List[torch.Tensor],
     inner_compile: Callable[..., Any] = compile_fx_inner,
     config_patches: Optional[Dict[str, Any]] = None,
-):
+) -> str:
```
Comment (angelayi) [inline torch/_inductor/compile_fx.py:852]: sorry, with https://github.com/pytorch/pytorch/pull/107468, this is still a callable.

File: torch/_inductor/codegen/wrapper.py [hunk 2]
```diff
@@ -1063,6 +1073,47 @@ def codegen_model_constructor(self):
...
+                    data_size = tensor.nelement() * tensor.element_size()
+                    self.prefix.writeline(
+                        f"constants_info_[{idx}].offset = {curr_offset};"
+                    )
+                    curr_offset += data_size
+                    self.prefix.writeline(
+                        f"constants_info_[{idx}].data_size = {data_size};"
+                    )
+                    if data_size == 0:
+                        V.graph.aot_constants.append(b"")
```
Comment (angelayi) [inline torch/_inductor/codegen/wrapper.py:1096]: Sorry maybe I don't understand the code too well here, but why are we appending to `V.graph.aot_constants` here as opposed to in graph.py, like where [`constants_repr` is updated](https://www.internalfb.com/code/fbsource/[a0c5bafcce08]/fbcode/caffe2/torch/_inductor/graph.py?lines=511)? It feels weird to be modifying `V.graph` in this file, which is just codegen-ing the wrapper.

**Review by angelayi — COMMENTED** [review]
AOT side generally looks fine to me! Do you mind including a paste of what the generated file looks like?

**Review by eellison — COMMENTED** [review]
Would you mind separating out the formatting changes or removing them ? (ghstack is nice for this)

File: torch/_inductor/codecache.py [hunk 3]
```diff
@@ -899,20 +937,61 @@ def compile(cls, graph, source_code, cuda):
...
+                    log.debug("aot compilation command: %s", cmd)
+                    run_command_and_check(cmd)
+
+                    consts_o = os.path.splitext(consts_path)[0] + ".o"
+                    cmd = f"ld -r -b binary -o {consts_o} {consts_path}"
+                    run_command_and_check(cmd)
+                    log.debug("aot constant binary command: %s", cmd)
+
+                    cmd = (
+                        "objcopy --rename-section"
```
Comment (chenyang78) [inline torch/_inductor/codecache.py:958]: ditto. Moreover, we may need to expose ```objcopy``` in a way similar to other binaries such as ```gcc``` and ```ld```. You can search ```make_build_paths``` in fbcode.

File: torch/_inductor/codecache.py [hunk 4]
```diff
@@ -899,20 +937,61 @@ def compile(cls, graph, source_code, cuda):
...
+                        f"objcopy --redefine-sym _binary_{body}_start=_binary_constants_bin_start {consts_o}"
+                    )
+                    symbol_list.append(
+                        f"objcopy --redefine-sym _binary_{body}_start=_binary_constants_bin_size {consts_o}"
+                    )
+                    symbol_list.append(
+                        f"objcopy --redefine-sym _binary_{body}_end=_binary_constants_bin_end {consts_o}"
+                    )
+                    log.debug(
+                        "aot constant binary redfine symbol: %s", " ".join(symbol_list)
```
Comment (chenyang78) [inline torch/_inductor/codecache.py:981]: typo ```redfine```?

File: torch/_inductor/codegen/wrapper.py [hunk 5]
```diff
@@ -1004,10 +1005,16 @@ def write_wrapper_decl(self):
                 isinstance(v, torch.Tensor) for v in list(V.graph.constants.values())
             ), "Expect all constants to be Tensor"
             for idx, constants_key in enumerate(V.graph.constants.keys()):
-                constants_idx = inputs_len + idx
-                self.prefix.writeline(
-                    f"at::Tensor {constants_key} = args[{constants_idx}];"
-                )
+                if V.graph.aot_mode:
+                    self.prefix.writeline(
+                        f"""at::Tensor {constants_key} = constants_->at("{constants_key}");"""
```
Comment (chenyang78) [inline torch/_inductor/codegen/wrapper.py:1010]: Is this called in the Model's ```run_impl``` method? If so, I am wondering if we could pre-set the key-to-index map so that we can have ```O(1)``` access to fetch a constant tensor, if it's not too hard. My worry is that the runtime overhead might be high if we had a large number of constants.

File: torch/_inductor/codegen/wrapper.py [hunk 6]
```diff
@@ -1062,6 +1072,32 @@ def codegen_model_constructor(self):
                         f"inputs_info_[{idx}].shape.emplace_back({size}, {size}, nullptr);"
                     )
 
+            constant_offset_mapping = dict(V.graph.aot_constant_offset)  # noqa: C416
```
Comment (chenyang78) [inline torch/_inductor/codegen/wrapper.py:1075]: Wondering why do we need ```# noqa: C416```?

File: torch/_inductor/graph.py [hunk 7]
```diff
@@ -208,6 +208,8 @@ def __init__(
         self.name = "GraphLowering"
         self.cpp_wrapper = cpp_wrapper
         self.aot_mode = aot_mode
+        self.aot_constant_curr_offset: int = 0
+        self.aot_constant_offset: List[Tuple[str, int]] = []
```
Comment (chenyang78) [inline torch/_inductor/graph.py:212]: Instead of attaching them in the general ```GraphLowering``` class, I am wondering if there would be another way to hold ```aot_constant_curr_offset``` and ```aot_constant_offset```. These two fields look too special to be placed here. 

Any thoughts or suggestion? @desertfire @eellison @jansel Thanks!

File: torch/_inductor/codegen/wrapper.py (hunk 5 above)
Comment (muchulee8) [reply torch/_inductor/codegen/wrapper.py:1010]: This is an `unordered_map`, so "in theory" it should be a O(1) access already. Key-to-index shouldn't matter much. as for not using `shared_ptr` this would be plenty of work I guess...

File: torch/_inductor/config.py [hunk 8]
```diff
@@ -74,6 +74,9 @@
 # If not specified, a temp directory will be created under the default caching path
 aot_inductor_output_path = ""
 
+# TODO: Temporary flag: If we are passing from export, ignore aot autograd for now
+ignore_aot_autograd = False
```
Comment (eellison) [inline torch/_inductor/config.py:78]: Let's move this and maybe `_in_aot_compilation` (my bad on that one) to `virtualized.py`, see: [this pr](https://github.com/pytorch/pytorch/pull/103289/files?show-viewed-files=true&file-filters%5B%5D=)  

adding `get_real_inputs` to `Virtualized.py`

File: torch/_inductor/freezing.py [hunk 9]
```diff
@@ -210,13 +210,19 @@ def freeze(
...
-        aot_autograd_gm, params_flat, fw_metadata
-    )
+        preserved_arg_indices = replace_params_with_constants(
+            aot_autograd_gm, params_flat, fw_metadata
+        )
+    else:
+        inputs = [
+            node for node in aot_autograd_gm.graph.nodes if node.op == "placeholder"
+        ]
+        preserved_arg_indices = list(range(len(inputs)))
```
Comment (eellison) [inline torch/_inductor/freezing.py:225]: hm, is this right, @angelayi had different changes here ?

File: torch/csrc/inductor/aot_inductor_model_container.h [hunk 10]
```diff
@@ -47,6 +69,46 @@ class AOTInductorModelContainer {
       output_names_.push_back(model->output_name(i));
       max_output_shapes_.emplace_back(model->max_output_shape(i));
     }
+
+    size_t num_constants = model->num_constants();
+    // Compute required blob size with 64-alignment
+    size_t max_blob = 0;
+    for (size_t i = 0; i < num_constants; i++) {
+      size_t data_size = model->constant_data_size(i);
+      if (data_size % 64) {
```
Comment (eellison) [inline torch/csrc/inductor/aot_inductor_model_container.h:78]: nit: would replace `64` with `alignment` constant

File: torch/_inductor/codegen/wrapper.py [hunk 11]
```diff
@@ -1062,6 +1072,32 @@ def codegen_model_constructor(self):
...
+            for idx, (name, tensor) in enumerate(V.graph.constants.items()):
+                assert isinstance(tensor, torch.Tensor)
+                self.prefix.writeline(f"""constants_info_[{idx}].name = "{name}";""")
+                self.prefix.writeline(
+                    f"constants_info_[{idx}].dtype = {DTYPE_TO_ATEN[tensor.dtype]};"
+                )
+                self.prefix.writeline(
+                    f"constants_info_[{idx}].offset = {constant_offset_mapping[name]};"
+                )
+                tensor = tensor.contiguous()
```
Comment (eellison) [inline torch/_inductor/codegen/wrapper.py:1085]: We should not be making tensors contiguous, this is going to silently give wrong with results with e.g. channels last constants. 

You can just serialize the raw storage, `tensor.untyped_storage()`, and then use the size, stride, storage_offset with the storage

File: torch/_inductor/graph.py [hunk 12]
```diff
@@ -505,17 +507,25 @@ def allocate():
...
+            if name is None:
+                name = f"constant{len(self.constants)}"
             self.constants[name] = data
             self.constant_reprs[name] = hashlib.sha256(
                 repr(data).encode("utf-8")
             ).hexdigest()
             return name
 
+        name = allocate(name)
+        tensor = data.contiguous()
```
Comment (eellison) [inline torch/_inductor/graph.py:521]: same thing here, we shouldnt need to call `.contiguous()`, and now the `tensor` has different  strides than `self.static_sizes_strides(data)`

File: test/inductor/test_inductor_freezing.py [hunk 13]
```diff
@@ -317,8 +317,8 @@ def foo(mod, x):
             # we unfuse the conv bias, but it should only have one constant in the kernel
             if self.device == "cuda":
                 FileCheck().check_not(".run(").check("conv").check(".run(").check_same(
-                    "constant"
-                ).check_not("constant").check_next("return").run(code[0])
+                    "frozen_param"
```
Comment (eellison) [inline test/inductor/test_inductor_freezing.py:320]: Why was this changed ? What is the error ?

**Review by eellison — COMMENTED** [review]
Cool ! did not review codecache. will defer to @desertfire on some of the code organization. 

It might be easier and more general to write the entire `untyped_storage()` of tensor constant, that would both handle striding and offsets, and allow you to dedup tensors with shared storages.

File: torch/_inductor/graph.py (hunk 7 above)
Comment (jansel) [reply torch/_inductor/graph.py:212]: Perhaps the wrapper code object?

File: torch/_inductor/codegen/wrapper.py (hunk 5 above)
Comment (muchulee8) [reply torch/_inductor/codegen/wrapper.py:1010]: This itself is an unordered_map, so we should have O(1) access to fetch the constant tensor. It is, possible to use a vector/list to hold the correct index. But if someday we want users to provide a mapping of constant name -> tensor, that's not possible for us to correctly render the index at compile time.

File: torch/_inductor/codegen/wrapper.py (hunk 6 above)
Comment (muchulee8) [reply torch/_inductor/codegen/wrapper.py:1075]: flake8 doesn't let me pass without this :(

File: torch/_inductor/graph.py (hunk 7 above)
Comment (muchulee8) [reply torch/_inductor/graph.py]: Done. I used `OrderedDict` instead so we could just iterate through the constant.

File: torch/csrc/inductor/aot_inductor_model_container.h [hunk 14]
```diff
@@ -47,6 +69,46 @@ class AOTInductorModelContainer {
...
+      constants_->emplace(
+          std::move(name),
+          at::from_blob(
+              constants_ptr, size, at::device(at::kCUDA).dtype(dtype)));
+
+      // Move GPU pointer to next aligned address.
+      if (data_size % 64) {
+        data_size = 64 + (data_size / 64) * 64;
+      }
+      constants_ptr = constants_ptr + data_size;
```
[the remark this answers is not in the document]
Comment (muchulee8) [reply torch/csrc/inductor/aot_inductor_model_container.h]: Done.

File: torch/csrc/inductor/aot_inductor_model_container.h (hunk 10 above)
Comment (muchulee8) [reply torch/csrc/inductor/aot_inductor_model_container.h]: Done

File: torch/_inductor/codecache.py [hunk 15]
```diff
@@ -899,20 +937,61 @@ def compile(cls, graph, source_code, cuda):
...
+                        vec_isa=picked_vec_isa,
+                        cuda=cuda,
+                        aot_mode=graph.aot_mode,
+                        compile_only=True,
+                    )
+                    log.debug("aot compilation command: %s", cmd)
+                    run_command_and_check(cmd)
+
+                    consts_o = os.path.splitext(consts_path)[0] + ".o"
+                    cmd = f"ld -r -b binary -o {consts_o} {consts_path}"
```
[the remark this answers is not in the document]
Comment (muchulee8) [reply torch/_inductor/codecache.py:997]: prerequisite: D48720058

File: torch/_inductor/freezing.py (hunk 9 above)
Comment (angelayi) [reply torch/_inductor/freezing.py:237]: Yup! Some other stuff changed, so I ended up updating this code, and then let Mu-Chu know about the updates.

Basically, torch.export recently added the `.module()` API which will unlift the parameters, and insert them as getattr nodes. In this case, instead of relying on inductor freezing to unlift them, we can just unlift it ahead of time and pass it directly to inductor. So, all the inputs can be preserved.

**Review by angelayi — APPROVED** [review]
frontend changes look fine to me!

File: torch/_inductor/config.py (hunk 8 above)
Comment (angelayi) [reply torch/_inductor/config.py]: We replaced this with a `from_export` config, which is just used to tell us that this is from export and to skip over AOTAutograd. I plan to remove this flag within the next 2 weeks after fixing some other stuff on the export side. So is it ok to add it here for now?

File: torch/_export/__init__.py [hunk 16]
```diff
@@ -553,11 +553,46 @@ def aot_compile(
...
-    param_buffer_values = list(ep.state_dict.values())
     flat_example_inputs = fx_pytree.tree_flatten_spec(
         combine_args_kwargs(args, kwargs), ep.call_spec.in_spec  # type: ignore[arg-type]
     )
-    all_args = (*param_buffer_values, *flat_example_inputs)
 
-    so_path = torch._inductor.aot_compile(ep.graph_module, list(all_args), options)
-    return so_path, ep
+    unlifted_module = ep.module()
+    unlifted_module.graph.set_codegen(torch.fx.CodeGen())  # type: ignore[attr-defined]
```
Comment (ipiszy) [inline torch/_export/__init__.py:561]: Wonder why is it necessary to do this? What's the CodeGen before this function call?

**Review by ipiszy — COMMENTED** [review]
Thanks @muchulee8 !

File: torch/_export/__init__.py (hunk 16 above)
Comment (angelayi) [reply torch/_export/__init__.py:561]: The CodeGen before this function call is a PytreeCodeGen, which I felt inductor didn't need to deal with at this time.

File: torch/csrc/inductor/aot_inductor_model_container.h [hunk 17]
```diff
@@ -6,6 +6,30 @@
...
+// We then turn the raw binary to an object file that exposes this
+// symbol and link it into the final .so.
+// For information on the binary format, see `man objcopy`, under
+// the "binary-architecture" flag:
+// https://man7.org/linux/man-pages/man1/objcopy.1.html
+// todo: use #embed in C++ 23 once available
+extern const uint8_t _binary_constants_bin_start[];
+extern const uint8_t _binary_constants_bin_end[];
+
+#define GPU_ALIGNMENT 64
```
Comment (eellison) [inline torch/csrc/inductor/aot_inductor_model_container.h:19]: This seems very likely to override some other constant elsewhere and does not need to be such a generic name or in the header

File: torch/csrc/inductor/aot_inductor_model_container.h [hunk 18]
```diff
@@ -51,6 +76,49 @@ class AOTInductorModelContainer {
...
+      // Create at::Tensor from copied memory.
+      auto dtype = model->constant_type(i);
+      auto size = model->constant_shape(i);
+      auto stride = model->constant_stride(i);
+      auto offset = model->constant_offset(i);
+
+      auto tensor = at::from_blob(
+          internal_ptr, size, stride, at::device(at::kCUDA).dtype(dtype));
+      tensor.unsafeGetTensorImpl()->set_sizes_and_strides(size, stride);
+      tensor.unsafeGetTensorImpl()->set_storage_offset(offset);
```
Comment (eellison) [inline torch/csrc/inductor/aot_inductor_model_container.h:119]: Where do we write out the tensors? We should add a test case with a storage offset

**Review by eellison — APPROVED** [review]

**Review by eellison — APPROVED** [review]
